### PR TITLE
Feature/provide requester secret refs in deal params

### PR DESCRIPTION
**Pull request by jbern0rd**
Ground work needed to handle requester secret references through `TaskDescription` in `iexec-sms`

### CODE REVIEW

File: src/main/java/com/iexec/common/task/TaskDescription.java [hunk 1]
```diff
@@ -142,6 +144,11 @@ public static TaskDescription toTaskDescription(String chainTaskId,
                 chainDeal.getParams().getIexecInputFiles() != null) {
             inputFiles = chainDeal.getParams().getIexecInputFiles();
         }
+        Map<String, String> secrets = Map.of();
+        if (chainDeal.getParams() != null &&
```
Comment (mcornaton) [inline src/main/java/com/iexec/common/task/TaskDescription.java]: Testing whether `chainDeal.getParams()` is null there is pretty strange as we use it a bunch of times without any caution later - [like here](https://github.com/iExecBlockchainComputing/iexec-common/blob/6273701ec7bc749f55d6f7cbe5b3c873f27d2aed/src/main/java/com/iexec/common/task/TaskDescription.java#L159).
So we're sure it'll throw an NPE if `chainDeal.getParams()` is null. What about adding a single test at the beginning of the method so that we're sure this value is never null later? BTW, that's already what's done for `chainDeal`.

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        DealParams params = DealParams.builder()
+                .iexecResultEncryption(false)
+                .iexecDeveloperLoggerEnabled(false)
+                .build();
+        Assertions.assertEquals("{}", RequestOrder.toStringParams(params));
+    }
+
+    @Test
+    void shouldSerializeParametersWithEncryptedResult() {
+        for (String storageProxy : List.of(DealParams.DROPBOX_RESULT_STORAGE_PROVIDER, DealParams.IPFS_RESULT_STORAGE_PROVIDER)) {
```
Comment (mcornaton) [inline src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: What about a parametrized test there, instead of a `for` loop?

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                .iexecInputFiles(List.of("file1", "file2"))
+                .build();
+        Assertions.assertEquals(expectedData, RequestOrder.toStringParams(params));
+    }
+
+    @Test
+    void shouldSerializeSecrets() {
+        String expectedData = "{\"iexec_secrets\":{\"0\":\"0\"}}";
+        DealParams params = DealParams.builder()
+                .iexecSecrets(Map.of("0", "0"))
```
Comment (mcornaton) [inline src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: I'd prefer to have two distinct values. Otherwise, there could be some switch between key and value and we could not notice that.

File: src/main/java/com/iexec/common/chain/DealParams.java [hunk 4]
```diff
@@ -50,6 +51,10 @@
     @JsonProperty("iexec_result_storage_proxy")
     private String iexecResultStorageProxy;
 
+    // Keys are app secrets indices, values are requester secrets indices
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/chain/DealParams.java:112]:
```suggestion
    /*
    The "iexec_secrets" key -> value map declares which secrets of the requester 
    can be accessed by a targeted TEE application.
    - A value should hold the secret name of a secret belonging to the requester.
    This secret must be previously pushed on the SMS. 
    - A key a associated to a value indicates how the TEE application can access 
    that value.
    
    Example:
    
    - A requester has a secret value "2000-01-01" pushed on the SMS. The name of 
    this secret is "my-birthday".
    - A requester has a secret value "+331234567890" pushed on the SMS. The name 
    of this secret is "my-phone-number".

    The application developer creates a TEE application which allows requesters 
    to :
    1. Share their birthday by affecting their secret to position "2"
    2. Share their phone number by affecting their secret to position "5"

    In order to share birthday and phone number to the TEE application, the 
    requester should declare:
    "iexec_secrets" : { "2": "my-birthday" , "5": "my-phone-number" }
    
    By doing this, the TEE application code will be able to read following 
    environment variables:
    IEXEC_REQUESTER_SECRET_2=2000-01-01
    IEXEC_REQUESTER_SECRET_5=+331234567890
    */
```

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java (hunk 3 above)
Comment (james-toussaint) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: Yes, plus a map having more than 1 mapping

File: src/main/java/com/iexec/common/chain/DealParams.java (hunk 4 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/common/chain/DealParams.java:112]: I agree with your suggestions, plus:
* I will try to add it as a javadoc comment
* Maybe the last part should be documented in `iexec-sms,` `iexec-common` has no knowledge of `IEXEC_REQUESTER_SECRET_` prefix

Comment (james-toussaint) [reply src/main/java/com/iexec/common/chain/DealParams.java:112]: Yes, I agree about responsibility of the last part. We can either :
- Add a note stating env var prefix is fake and just for overall understandability purposes
- Relocate that part to `com.iexec.common.utils.IexecEnvUtils`

Comment (zguesmi) [reply src/main/java/com/iexec/common/chain/DealParams.java:112]: I think it's a good idea to add `IEXEC_REQUESTER_SECRET_` to `IexecEnvUtils`.

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package com.iexec.common.sdk.order.payload;
+
+import com.iexec.common.chain.DealParams;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+import java.util.Map;
+
+class RequestOrderTest {
```
Comment (zguesmi) [inline src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java:7]: Can we add a test that serializes all params at once?

File: src/main/java/com/iexec/common/task/TaskDescription.java (hunk 1 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/common/task/TaskDescription.java]: It would be possible if `chainDeal.getParams()` was the only call which could throw a NPE.
NPE can be thrown by the following calls: `getBotSize()`, `getChainApp()`, `getChainCategory()`, `getChainDataset()`
It is not possible to do a single test at the beginning with those multiple calls.
**I will remove the unnecessary check and add a TODO.**

Later, I would like to investigate how data read from the chain are populated in Object instances.
It would be great to guarantee that `chainDeal.getParams()` can never be null.

What does everyone think of this suggestion ?

Comment (zguesmi) [reply src/main/java/com/iexec/common/task/TaskDescription.java]: Actually `chainDeal.getParams()` cannot be null. See how it is created in the deal [here](https://github.com/iExecBlockchainComputing/iexec-common/blob/63f66080c6ab269897a7d3bd125479b3b6b0eef1/src/main/java/com/iexec/common/chain/IexecHubAbstractService.java#L403). Cf [here](https://github.com/iExecBlockchainComputing/iexec-common/blob/63f66080c6ab269897a7d3bd125479b3b6b0eef1/src/main/java/com/iexec/common/chain/ChainDeal.java#L71).

Comment (zguesmi) [reply src/main/java/com/iexec/common/task/TaskDescription.java]: Also just for information, the deal params field onchain is not guaranteed to be non-null. So the object `DealParams` will be defined but the content of [iexecArgs](https://github.com/iExecBlockchainComputing/iexec-common/blob/63f66080c6ab269897a7d3bd125479b3b6b0eef1/src/main/java/com/iexec/common/chain/ChainDeal.java#L81) could potentially be null which is not necessarily a problem.

Comment (jbern0rd) [reply src/main/java/com/iexec/common/task/TaskDescription.java]: I will take another look to improve my understanding of the code.
Thanks for the help.

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java (hunk 2 above)
Comment (jbern0rd) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/7107a8db754fd4807bf1ec149afd78fa41506bd1

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java (hunk 3 above)
Comment (jbern0rd) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/7107a8db754fd4807bf1ec149afd78fa41506bd1

File: src/main/java/com/iexec/common/chain/ChainDeal.java [hunk 6]
```diff
@@ -72,15 +73,19 @@ public static DealParams stringToDealParams(String params) {
...
+            if (dealParams.getIexecSecrets() == null) {
+                dealParams.setIexecSecrets(Map.of());
             }
             return dealParams;
         } catch (IOException e) {
             //the requester want to execute one task with the whole string
             return DealParams.builder()
                     .iexecArgs(params)
-                    .iexecInputFiles(new ArrayList<>())
+                    .iexecInputFiles(List.of())
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/chain/ChainDeal.java]:
```suggestion
                    .iexecInputFiles(Collections.emptyList())
```
?
https://stackoverflow.com/a/39400310

File: src/main/java/com/iexec/common/chain/ChainDeal.java [hunk 7]
```diff
@@ -72,15 +73,19 @@ public static DealParams stringToDealParams(String params) {
...
                     .iexecDeveloperLoggerEnabled(false)
+                    .iexecSecrets(Map.of())
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/chain/ChainDeal.java]:
```suggestion
                    .iexecSecrets(Collections.emptyMap())
```

File: build.gradle [hunk 8]
```diff
@@ -61,9 +61,7 @@ dependencies {
     implementation 'net.jodah:failsafe:2.4.4'
 
     // test
-    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.6.0'
-    testImplementation 'org.junit.jupiter:junit-jupiter-engine:5.6.0'
-    testImplementation 'org.mockito:mockito-core:3.6.0'
+    testImplementation 'org.junit.jupiter:junit-jupiter:5.6.0'
```
Comment (james-toussaint) [inline build.gradle:75]: :+1:

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java (hunk 5 above)
Comment (jbern0rd) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java:7]: Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/7107a8db754fd4807bf1ec149afd78fa41506bd1

File: src/test/java/com/iexec/common/chain/ChainDealTest.java [hunk 9]
```diff
@@ -62,26 +63,57 @@
...
+        Assertions.assertNull(deal.getParams());
+        Assertions.assertNull(deal.getPoolOwner());
+        Assertions.assertNull(deal.getPoolPointer());
+        Assertions.assertNull(deal.getPoolPrice());
+        Assertions.assertNull(deal.getRequester());
+        Assertions.assertNull(deal.getSchedulerRewardRatio());
+        Assertions.assertNull(deal.getStartTime());
+        Assertions.assertNull(deal.getTag());
+        Assertions.assertNull(deal.getTrust());
+        Assertions.assertNull(deal.getWorkerStake());
```
Comment (zguesmi) [inline src/test/java/com/iexec/common/chain/ChainDealTest.java]: You could use `hasAllNullFieldsOrProperties()` from `org.assertj.core.api.Assertions.assertThat`.

File: src/test/java/com/iexec/common/chain/ChainDealTest.java [hunk 10]
```diff
@@ -62,26 +63,57 @@
...
+    }
+
     @Test
     void shouldReadArgsWithoutJson() {
         DealParams params = stringToDealParams(ARGS);
         assertEquals(ARGS, params.getIexecArgs());
         assertEquals(0, params.getIexecInputFiles().size());
+        assertEquals(0,params.getIexecSecrets().size());
```
Comment (zguesmi) [inline src/test/java/com/iexec/common/chain/ChainDealTest.java]:
```suggestion
        assertEquals(0, params.getIexecSecrets().size());
```

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java [hunk 11]
```diff
@@ -48,18 +63,47 @@ void shouldSerializeArgs() {
...
         Assertions.assertEquals(expectedData, RequestOrder.toStringParams(params));
     }
 
     @Test
     void shouldSerializeSecrets() {
-        String expectedData = "{\"iexec_secrets\":{\"0\":\"0\"}}";
+        Map<String, String> secrets = Map.of("0", "secretA", "2", "secretD");
+        String serializedSecrets = secrets.entrySet().stream()
+                .map(entry -> "\"" + entry.getKey() + "\":\"" + entry.getValue() + "\"")
+                .collect(Collectors.joining(","));
```
Comment (zguesmi) [inline src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: I find this more readable:
```suggestion
        String serializedSecrets = "{\"0\":\"secretA\", \"2\":\"secretD\"}"
```

Comment (jbern0rd) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: Sadly, I don't have much of a choice here.
The keys are unordered in a Map.
Serialized data can thus be `{"0":"secretA","2":"secretD"}` or `{"2":"secretD","0":"secretA"}`
That's why I decided to perform my own serialization with a stream...

Comment (zguesmi) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: I see. For that reason I generally prefer reading `expectedData` in a `JSONObject` and compare its fields.
No problem, we can keep it that way.

Comment (mcornaton) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: What if you provide a `LinkedHashMap` instead? Its order should be predictable.
See https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html.

File: src/main/java/com/iexec/common/chain/DealParams.java (hunk 4 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/common/chain/DealParams.java:112]: I have a question about this.
Do I just add the String definition to `IexecEnvUtils` ?
`public static final String IEXEC_REQUESTER_SECRET_PREFIX = "IEXEC_REQUESTER_SECRET_";`
Do I add `IEXEC_APP_DEVELOPER_SECRET_PREFIX` as well ?

File: src/main/java/com/iexec/common/chain/ChainDeal.java (hunk 6 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/common/chain/ChainDeal.java]: Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/c642ed8631c33d38f36a6e45da21850c8272bf56

File: src/main/java/com/iexec/common/chain/ChainDeal.java (hunk 7 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/common/chain/ChainDeal.java]: Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/c642ed8631c33d38f36a6e45da21850c8272bf56

File: src/test/java/com/iexec/common/chain/ChainDealTest.java (hunk 9 above)
Comment (jbern0rd) [reply src/test/java/com/iexec/common/chain/ChainDealTest.java]: Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/86b25b4e02c469cd822bf7e25a2dfd36e10c0e5b

File: src/test/java/com/iexec/common/chain/ChainDealTest.java (hunk 10 above)
Comment (jbern0rd) [reply src/test/java/com/iexec/common/chain/ChainDealTest.java]: Fixed in https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/86b25b4e02c469cd822bf7e25a2dfd36e10c0e5b

File: src/main/java/com/iexec/common/task/TaskDescription.java (hunk 1 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/common/task/TaskDescription.java]: See https://github.com/iExecBlockchainComputing/iexec-common/pull/298/commits/f3b818be8003c042ebf381f70553e7290a43b315

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java (hunk 11 above)
Comment (jbern0rd) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: I think what we really care is proving something like:
`stringToDealParams(toStringParams(dealParams)).equals(dealParams)` but is not the same java Object

`ChainDeal#toStringParams` could be refactored to `DealParams#toString`
`RequestOrder#toStringParams` could be refactored to `DealParams#parseString`

To provide the maximum value, I would prefer to do this kind of refactoring instead of trying to check the correct serialization and deserialization of hard-coded json strings.

File: src/main/java/com/iexec/common/chain/DealParams.java (hunk 4 above)
Comment (zguesmi) [reply src/main/java/com/iexec/common/chain/DealParams.java:112]: We can add both yes. Even though the variables will probably be used only in the SMS, it's a good idea to have them all at the same place.

**Review by zguesmi — APPROVED** [review]
Well done!

**Review by mcornaton — APPROVED** [review]
LGTM! :)

File: src/main/java/com/iexec/common/chain/DealParams.java (hunk 4 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/common/chain/DealParams.java:112]: See 23794c6  and 7040309

File: src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java (hunk 11 above)
Comment (jbern0rd) [reply src/test/java/com/iexec/common/sdk/order/payload/RequestOrderTest.java]: See 23794c6

**Review by zguesmi — APPROVED** [review]
Clean!

File: src/main/java/com/iexec/common/chain/DealParams.java (hunk 4 above)
Comment (james-toussaint) [reply src/main/java/com/iexec/common/chain/DealParams.java:112]: Thanks!

File: src/main/java/com/iexec/common/chain/DealParams.java [hunk 12]
```diff
@@ -58,5 +121,48 @@
...
 
+    /**
+     * Creates an instance from a JSON string representation.
+     * @param paramString data to create a DealParams instance.
+     *                    The value will be parsed if it is a valid JSON string,
+     *                    it will be used as <b>iexecArgs</b> value otherwise.
+     *                    In the second case, every other parameters will have default values.
+     * @return the created instance
+     */
+    public static DealParams createFromString(String paramString) {
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/chain/DealParams.java:132]: :+1:

File: src/test/java/com/iexec/common/chain/DealParamsTest.java [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+                .iexecInputFiles(List.of(FILE2, FILE3))
+                .iexecSecrets(Map.of("2", "secretK", "1", "secretD"))
+                .build();
+        DealParams newParams = DealParams.createFromString(params.toJsonString());
+        assertThat(params).isEqualTo(newParams);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {DealParams.DROPBOX_RESULT_STORAGE_PROVIDER, DealParams.IPFS_RESULT_STORAGE_PROVIDER})
+    void testSerializationWithAllParams(String storageProvider) {
```
Comment (james-toussaint) [inline src/test/java/com/iexec/common/chain/DealParamsTest.java:172]: :+1:

**Review by james-toussaint — APPROVED** [review]
Thanks :)

File: src/main/java/com/iexec/common/chain/DealParams.java [hunk 14]
```diff
@@ -17,11 +17,69 @@
...
 import java.util.Map;
 
+import static com.fasterxml.jackson.annotation.JsonInclude.Include.*;
+
+/**
+ * The DealParams class represents runtime configurations provided for tasks computation.
+ * DealParams allow a requester to provide <b>args</b>, <b>input files</b>, <b>secrets</b>,
+ * <b>result storage configuration</b>... Those configurations are written on the blockchain
+ * when a deal is submitted.
+ * <p>
```
Comment (mcornaton) [inline src/main/java/com/iexec/common/chain/DealParams.java:38]: I usually prefer to use `<br>` tag. Don't know what @jeremyjams and @zguesmi prefer.

**Review by mcornaton — APPROVED** [review]
Looks great, thanks!

File: src/main/java/com/iexec/common/chain/DealParams.java (hunk 14 above)
Comment (james-toussaint) [reply src/main/java/com/iexec/common/chain/DealParams.java:38]: b stands for bold :)

Comment (jbern0rd) [reply src/main/java/com/iexec/common/chain/DealParams.java:38]: I followed this documentation https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html
I like `<p>` which stands for paragraph but `<br>` is fine as well.

Comment (james-toussaint) [reply src/main/java/com/iexec/common/chain/DealParams.java:38]: OK comment was on `<p>` ..!

Default Intellij likes `<p>` too:
image

Comment (mcornaton) [reply src/main/java/com/iexec/common/chain/DealParams.java:38]: If both of you prefer `<p>`, let's use that tag so we're consistent :)
